Property Graph: multiple topologies

amberhassaan opened the pull request:
This is the first implementation of multiple-topologies vision for PropertyGraph. The idea is to be able to generate multiple topologies on demand, such as sorted, or transposed. Thus exposing additional functionality.

Wait for https://github.com/KatanaGraph/katana-enterprise/pull/1537

bowu commented on libquery/src/MatchOp.cpp:
```diff
@@ -12,8 +13,22 @@ MatchOp::MatchOp(const galois::BitLabelSchema& bit_label_schema,
+  // passed into the querying engine
+  std::vector<uint32_t> query_vertex_to_matching_order;
+
+  QueryEmbeddingQueueType matched_embeddings = data_graph_->MatchQueryGraph(
```
We can discuss later whether MatchQueryGraph should be a member function of data graph. It should work for now.

amberhassaan committed:
After squashing and rebasing on master

WIP: initial skeleton for labeled topology classes for PropertyGraph

WIP: adding more functionality to LabeledTopology

WIP: checkpoint. wrapped topology APIs in PropertyGraph

WIP: separate raw transpose topology

remove duplicate functions

WIP: checkpoint Added zip iterator. cleaning up label sorted topology

WIP: Completed bare minimum implementation of topology code

WIP: add bi-directional label-aware topology

WIP: Fix build. Add overload to katana::Iterate()

Fix build, bugs and format

Fixed bug re: creating label aware topology from transpose topology

Move long methods to .cpp

Fix Lint error

fix lint error

Add SFINAE to help cython pick the right overload of iterate

fixed a bug with SortEdgesByDest

Fix a corner condition

Bug fix to assert with bad condition

Fixed another corner case with 0 edges

Restrict topology interface. Return empty range instead of
std::optional. Replace edge label with edge type in names

fix build

fix unused variable warning

fix warning

gcc complains while syntax is valid

fix array indexing bug

remove edge_begin/edge_end

added a test for zip iterator

fixed constant naming style

build fix for gcc-7

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -29,18 +29,20 @@ ProjectAsArrowArray(const T* buf, const size_t len) noexcept {
-/// format.
-class KATANA_EXPORT GraphTopology {
-public:
+struct KATANA_EXPORT GraphTopologyTypes {
```
What's the value of having this in a separate container struct?

amberhassaan replied:
Other topologies inherit this to get the typedefs.

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -125,6 +127,13 @@ class KATANA_EXPORT GraphTopology {
     return MakeStandardRange<node_iterator>(begin, end);
   }
 
+  nodes_range all_nodes() const noexcept {
```
Why not just `nodes`?

amberhassaan replied:
My intention is to clarify with the name that you will iterate over all nodes. There is nodes(beg, end) and I wanted to differentiate from that instead of overload. Similarly edges(node_id) vs edges() (as in all_edges()) are different, so I didn't want them to get confused with each other.

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -135,6 +144,49 @@ class KATANA_EXPORT GraphTopology {
+    return std::distance(edge_range.begin(), edge_range.end());
+  }
+
+  Edge original_edge_id(const Edge& eid) const noexcept { return eid; }
+
+  Node original_node_id(const Node& nid) const noexcept { return nid; }
```
These need to be documented. "Original" could mean a lot of things.

amberhassaan replied:
Yes, I agree. My other choice was to introduce "PropertyIndex", but then I realized that these are in fact NodeIDs/EdgeIDs, except that they are original, unsorted NodeIDs.

roshandathathri replied:
This is quite confusing. This is NOT a `LocalNodeID` or `LocalEdgeID` right? 
I'd prefer a `OpaqueID` for `NodePropertyID` and `EdgePropertyID`. This will also handle grouping of properties (in the future).

amberhassaan replied:
We can go two ways here (adding @scober for comments too as he mentioned property_index):
1) We introduce PropertyIndex concept and these methods become `node_property_index(Node)` and `edge_property_index(Edge)` where the node/edge ID passed is the renumbered ID (and not the original ID). All properties are accessed through this property_index. (Note that this will change how `GetData/GetEdgeData` works for TypedPropertyGraph). Also, say that we implement grouping of Properties, you can imagine that we'll keep some map from Node IDs to property index. Now, all topologies will keep their own map for their renumbered Node/EdgeIDs to PropertyIndex, as opposed to a central map in option 2. 
 
2) We establish the invariant that properties can always be accessed with just Node/Edge ID and optional topologies have the methods above to obtain the unsorted Node/Edge IDs.

scober replied:
I like option 1 for the understandability reasons mentioned by Roshan and Arthur.

I would like to push that suggestion to include the base `PropertyGraph`, though. So that even if you have a "regular" `PropertyGraph` and not a view, you still need to call `{node,edge}_property_index`.

That would bring the interfaces closer together and give the sense that the base `PropertyGraph` is still a view, it is just the default view. I also like the idea that the partitioner could create property arrays + an indirection layer for the base `PropertyGraph` instead of a property table ordered by node/edge ids.

arthurp replied:
I don't fully understand option 2 here. But I think option 1 sounds good. And it must be used in both the views and the main PropertyGraph for consistency.

roshandathathri replied:
I'm in favor of (1). That fits well into our long term plans.

amberhassaan replied:
Hi guys, I know we all like option (1) but I'm a bit hesitant. Here's why. With option (1), every topology maintains its own map from Node/Edge IDs to PropertyIndex. Additionally, topologies that are Label/Type aware will need to copy the mapping from Node/Edge IDs to TypeSetIDs and rearrange it upon sorting, just like the PropertyIndex. 

Now, with option (2), we don't copy any of these mappings. There's only mapping from original (unsorted) Node/Edge IDs to PropertyIndex, and similarly for TypeSetIDs. The optional topologies keep a mapping from sorted/rearranged Node/Edge IDs to original Node/Edge IDs. So a PropertyIndex (or TypeSetID) lookup goes as sorted Node ID ---> unsorted ID ---> PropertyIndex.  In a way, option(1) is flattening this level of indirection to sorted Node ID ---> PropertyIndex, at the cost of extra memory (and extra swaps). 

I'm unsure at the moment and tempted to leave it as is until we implement grouping of properties and have non-trivial PropertyIndices. Thoughts?

roshandathathri replied:
Thanks. I didn't understand the extra memory overhead. Why would the overhead be different? Every topology, except original, will need a map from node ID -> property index or unsorted node ID, right?

Also, original topology will also need mapping from node ID -> property index once we compact properties. If the mapping is identity, then we can specialize for that.

arthurp replied:
I think we are mixing API and implementation here. I'm fine with having two levels of indirection. That's not problem. The important this is that in the API we do not require that the user does the mapping themselves. And I like the API implied by (1). So if we can implement that API using the more compact set of tables, I would we perfectly happy.

amberhassaan replied:
OK, I hear you guys and I agree, but I'm going to postpone this change to my next PR that will implement the PropertyGraph Views. I'd like to close this PR without introducing more changes (apart from addressing the comments). 

@roshandathathri : there is a minor difference in overheads between options (1) and (2), and it may not matter. With (1), we need to copy typeID arrays and sort them, but on the plus side, we remove one level of indirection in look ups. With (2), we only need to maintain sorted ID --> unsorted ID mapping, so we sort only the ID array (in addition to destinations array). 

There is one point to ponder. Having topologies maintain property_index for lookups means that we need to define the interface between topologies and the (non-existent as of now) Property Store. Property Store is what will remain of PropetyGraph once we remove the topology related guts from it and leave only Property related code. We will still have PropertyGraph which will be the view with default (original/unsorted) topology.

arthurp replied:
OK. I get that. Thanks for cutting off the bike shedding.

My specific request is to document the API **in this PR** so it's really clear which type of ID is expected by and returned by each method. Is it an original/default ID? Is it a ID for the sorted view? For another view? That way we can a) correctly use this API while it exists and b) have a good way to talk about the various types of IDs while we hide them behind abstractions.

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -135,6 +144,49 @@ class KATANA_EXPORT GraphTopology {
+
+  Node original_node_id(const Node& nid) const noexcept { return nid; }
+
+#if 0
```
Either remove or implement these before merge.

amberhassaan replied:
Will do.

amberhassaan replied:
Leaning towards removing these as the `local_ranges` are not well defined for NUMAArray (esp. with Interleaved allocation policy).

amberhassaan replied:
remove local_ranges.

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -767,6 +819,620 @@ CreateSymmetricGraph(PropertyGraph* pg);
 CreateTransposeGraphTopology(const GraphTopology& topology);
 
+// TODO(amber): add NodeLabelVecT
+// template <typename NodeLabelVecT, typename EdgeLabelVecT>
```
Will this be part of the public API? We are trying to keep templates out of the public untyped API. It makes Python integration much easier and provides a nice clean separation between organization of data and the specific data interpretation.

amberhassaan replied:
No. This is leftovers from earlier WIP. Sorry about that.
That said, I might introduce templates in such a way that  they don't make their way to Python. The issue at hand is that I don't want dynamic polymorphism/virtual methods for Graph topology since these are short frequent function calls. I might use static polymorphism via templates. In the end, you will get family of topologies and correspondingly a family of PropertyGraph "views". I'll update the RFC and we can discuss this further there. (Thanks for the comments on RFC, I'll respond to those as well).

arthurp replied:
I don't fully understand what you mean here. But I see your general point. One option might be to use static polymorphism in C++, but still have the virtual functions so we don't need to create a bunch of different cases and potentially have combinatorial explosion in the Python wrappers. I did some testing and if you mark the subclass methods as final then they can be inlined just like a non-virtual function if you have a reference to that time directly. See: https://github.com/KatanaGraph/katana-enterprise/blob/69e3de6fb7c6878e8e9024b2aff6956aee96b25d/libdistgalois/include/katana/WorkDetection.h

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -767,6 +819,620 @@ CreateSymmetricGraph(PropertyGraph* pg);
+
+class KATANA_EXPORT EdgeLabelAwareTopology;
+
+// TODO(amber): make OrigEdgeIDMap optional via template argument
```
I want to be really careful with this. Configuration via template is very hard to wrap from Python. That said if there are a few common choices and the objects are always passed around via pointer or reference (using dynamic typing) then it would be OK.

amberhassaan replied:
see above.

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -767,6 +819,620 @@ CreateSymmetricGraph(PropertyGraph* pg);
+public:
+  TransposeTopology() = default;
+  TransposeTopology(TransposeTopology&&) = default;
+  TransposeTopology& operator=(TransposeTopology&&) = default;
```
Don't move semantics not play nice with subclassing? I feel like we might need to use pointers and references to allow functions to take an arbitrary `GraphTopology` without a template parameter. Which is important for exposing this to Python. Sorry to repeat myself.

amberhassaan replied:
I don't see why move semantics won't work with simple inheritance.

arthurp replied:
Because you cannot move a value of type `TransposeTopology` into a variable of type `GraphTopology`. This would prevent code that was not templatized from performing moves abstractly. There may be a way around this, so that non-templatized code never needs to perform a move, but it seems like it could cause problems.

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -767,6 +819,620 @@ CreateSymmetricGraph(PropertyGraph* pg);
+  }
+
+  Edge original_edge_id_using_in_edge(
+      const edge_iterator& in_edge_iter) const noexcept {
```
I'd like to get ride of the edge_iterator variants of these if possible. Having two versions just doesn't make much sense. And the iterator version is the one people use when they are NOT using range-for and I want to encourage people to use range-for.

amberhassaan replied:
Can do.

amberhassaan replied:
done.

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -767,6 +819,620 @@ CreateSymmetricGraph(PropertyGraph* pg);
+   */
+  std::optional<edges_range> FindAllEdges(Node src, Node dst) const {
+    if (OutTopoBase::degree(src) == 0 || in_topo_.degree(dst) == 0) {
+      return std::nullopt;
```
Why not return an empty range?

amberhassaan replied:
Let me think about that. I lifted these functions from LC_CSR_CSC_Labeled_Graph, which was being used by libquery's AttributedGraph.

roshandathathri replied:
@aneeshdurg Can you please check if libquery is fine to return an empty range for all these `Find*` functions?

amberhassaan replied:
we have empty range now.

arthurp reviewed (changes requested):
I added some comments, but ran out of time. I'll get back to this tomorrow or late tonight.

I think we should discuss this API. You seem to be planning  to use templates and I don't think that's a good idea.

ddn0 commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -136,6 +149,54 @@ class KATANA_EXPORT GraphTopology {
+  friend class EdgeShuffleTopology;
+  friend class EdgeLabelAwareTopology;
+
+  NUMAArray<Edge>&& MoveAdjIndices() noexcept {
```
I would prefer that these functions are just accessors like `NUMAArray<Edge>& adj_indices()` without an additional variant for rvalues as any lvalue can be turned to an rvalue. Callers can use `std::move()` themselves.

amberhassaan replied:
done.

ddn0 commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +903,633 @@ CreateSymmetricGraph(PropertyGraph* pg);
+// template <typename NodeLabelVecT, typename EdgeLabelVecT>
+
+class KATANA_EXPORT EdgeLabelIndex : public GraphTopologyTypes {
+  using EdgeLabelTy = PropertyGraph::TypeSetID;
```
This type is referenced in the public API via `label`. So, it would be nice for this to be part of the public API as well.

Can you call this something like `EdgeLabel`? The `Ty` suffix is confusing to most as these are just types just like any other type.

amberhassaan replied:
done.

ddn0 commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +903,633 @@ CreateSymmetricGraph(PropertyGraph* pg);
+   * Returns all edges from src to dst with some label.  If not found, returns
+   * nothing.
+   */
+  std::optional<edges_range> FindAllEdgesWithLabel(
```
return empty range instead of `optional`?

amberhassaan replied:
done.

ddn0 commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +903,633 @@ CreateSymmetricGraph(PropertyGraph* pg);
+  // orig_edge_ids_
+
+protected:
+  using EdgeLabelTy = PropertyGraph::TypeSetID;
```
This is part of the public API via `edge_begin` this should be public as well. Otherwise, callers cannot provide explicit types to the arguments to public API functions.

amberhassaan replied:
done.

ddn0 commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +903,633 @@ CreateSymmetricGraph(PropertyGraph* pg);
+
+  // Edge accessors
+
+  /**
```
can we stick with one style of Doxygen comments in a file? `///`

amberhassaan replied:
done

ddn0 commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +903,633 @@ CreateSymmetricGraph(PropertyGraph* pg);
+  OrigEdgeIDMap orig_edge_ids_;
+};
+
+/**
```
`///`

amberhassaan replied:
done

ddn0 commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +903,633 @@ CreateSymmetricGraph(PropertyGraph* pg);
+    : public EdgeLabelAwareTopology {
+  // Inheriting to reuse outgoing topology API
+  //
+  using OutTopoBase = EdgeLabelAwareTopology;
```
unless you need to refer to types in a parent class, you can just use `this->foo` to refer to functions in your parent class.

ddn0 commented on libsupport/include/katana/Iterators.h:
```diff
@@ -0,0 +1,187 @@
+ * because either left or right parameter to the comparator can be a
+ * tuple of references or tuple of values depending on the calling context.
+
+int main() {
```
I don't know if Doxygen will like this formatting.

I suggest just using `///` on every line.

amberhassaan replied:
done

ddn0 commented on libsupport/include/katana/Iterators.h:
```diff
@@ -0,0 +1,187 @@
+ * overload or conflict with std::swap for std::tuple<>
+ */
+template <typename... Args>
+class KATANA_EXPORT ZipRefTuple : public std::tuple<Args&...> {
```
Can you add a test?

amberhassaan replied:
done

ddn0 commented on libsupport/include/katana/Iterators.h:
```diff
@@ -0,0 +1,187 @@
+*/
+
+/**
+ * This is a thin wrapper around std::tuple<Args&...> . It is needed to
```
Can you explain why this is different than https://www.boost.org/doc/libs/1_41_0/libs/iterator/doc/zip_iterator.html ?

amberhassaan replied:
I was seeing errors due to ill-defined `swap` originating from `libstdc++` code. I can reproduce a small example for you. 

The other issue is that `boost::tuple` has a different `get<>()` syntax from `std::tuple` and I chose to prefer `std::tuple` syntax. It's simpler. boost syntax requires things like `tup.template get<0>()`.

amberhassaan replied:
fyi: I started with `boost::zip_iterator` before going down this rabbit hole and ruining my Saturday (last weekend). I suspect there might be a way to fix `boost::zip_iterator` but then I don't like the `boost::tuple` syntax.

roshandathathri replied:
Is this the same as `boost::combine`?

amberhassaan replied:
Some similarities. We'll have to test it against `std::sort()` to see if it really works. `std::sort()` implementation is a good test because it performs `swap` and move/copy assignment in its code.

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -30,18 +31,23 @@ ProjectAsArrowArray(const T* buf, const size_t len) noexcept {
+};
+
+class KATANA_EXPORT EdgeShuffleTopology;
+class KATANA_EXPORT EdgeLabelAwareTopology;
```
"Label" and "Type" are confusing. I don't think we resolved or came to a consensus on type, label, or tag - but we should be consistent for now. I'd prefer if we replaced "Label" with "Type".

In some places, I'm assuming that by "Label", you mean "TypeSetID" in PropertyGraph? In such cases, I'd prefer to explicitly say `TypeSetID`.

amberhassaan replied:
done

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -136,6 +149,54 @@ class KATANA_EXPORT GraphTopology {
+
+private:
+  friend class EdgeShuffleTopology;
+  friend class EdgeLabelAwareTopology;
```
Can we avoid friend classes? Why is this needed? Please add comments in the code as well.

amberhassaan replied:
added comment. Can't avoid friend classes for now.

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +903,633 @@ CreateSymmetricGraph(PropertyGraph* pg);
+// TODO(amber): add NodeLabelVecT
+// template <typename NodeLabelVecT, typename EdgeLabelVecT>
+
+class KATANA_EXPORT EdgeLabelIndex : public GraphTopologyTypes {
```
Ditto about label vs. type

amberhassaan replied:
done.

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +903,633 @@ CreateSymmetricGraph(PropertyGraph* pg);
+  uint32_t index(const EdgeLabelTy& label) const noexcept {
+    KATANA_LOG_ASSERT(edge_label_to_index_map_.count(label) > 0);
+    // return edge_label_to_index_map_[label];
+    return edge_label_to_index_map_.find(label)->second;
```
I think this is outdated code. Both these maps should have (key == value) pairs. The way we constructed `TypeSetID` earlier was that they could be non-contiguous. Here, they are contiguous. 

There is still a case where it makes sense to have these maps - when there are no edges of a particular `TypeSetID`, we can skip that and save O(V) space in the out-index. But I don't think it's worth it right now. Please add a TODO and drop the maps. It'll make the code much simpler.

---
On second thoughts, you can leave the code as-is and add a TODO to drop the maps. It'll simplify the impact of this PR. I leave the choice up to you.

roshandathathri replied:
Third thoughts: There could be cases where this is useful. So it's fine to leave it as-is. But change the names to "type" instead of "label".

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +903,633 @@ CreateSymmetricGraph(PropertyGraph* pg);
+        return true;
+      }
+    }
+    return false;
```
nit: I'm a bit wary of all function definitions being in the header. As it's not templated, some of these could be moved to the CPP file.

amberhassaan replied:
Method definitions inside the class are considered `inline` and guaranteed to not generate `duplicate definition` error. So, it's a matter of judgement to move large methods to .cpp if inlining is not expected to help.

roshandathathri replied:
Yes, I'm not concerned about duplicate definitions. I'm concerned about increasing in compilation time.

roshandathathri commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -1111,3 +1111,239 @@ katana::CreateTransposeGraphTopology(const GraphTopology& topology) {
+    edge_index_to_label.emplace_back(edgeLabelPair.first);
+  }
+
+  // TODO(amber): introduce a per-thread-container type that frees memory
```
I think this should be handled by `PerThreadStorage` destructor transparently.

amberhassaan replied:
I have put a todo that will take care of most of these cases. I don't want to change PerThreadStorage destructor as that will apply to every PerThreadStorage object.

roshandathathri replied:
My point is that it should apply for every PerThreadStorage object. Such objects will only be deallocated on the main thread always and all the memory will go to that thread's pool. This is bad for non-POD types.

amberhassaan replied:
I think you're confusing PerThreadStorage with PerThreadAllocator. A non-trivial/non-POD PerThreadStorage object may not use a per-thread allocator. However, gstl style containers when used with PerThreadStorage are a case that uses Per-Thread-Allocator too. So my suggestion (& plan) is to implement a generic "Per-thread-container" for these situations.

roshandathathri replied:
Oh.. right! If the non-POD type is NOT `gstl` style container, then there is no problem. 

Yes, that makes sense. Thanks.

arthurp commented:
As written this API will require some significant duplication of code to wrap it for python because the types don't allow dynamic polymorphism. This will mean I will need to separately wrap each type and cannot share wrappers for any methods of GraphTopology which are overridden in the subclasses.

There is no need to change this now. I just wanted to bring it up so people can comment and correct sooner rather than later if needed. When I go to wrap this, I expect I will extract the topology interface into a superclass that will have the overridden methods be virtual. This will allow me to use the constructors in the views and move the returned topology into a unique_ptr which can be used as either a GraphTopology or as the appropriate subclass by the Python wrappers. It should be possible to avoid any overhead other than the vtable pointer in the representation in C++ code that directly names the implementation classes.

arthurp commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -849,30 +856,24 @@ katana::GraphTopology::Edge
+  auto e_range = topo.edges(src);
 
-  constexpr int BINARY_SEARCH_THRESHOLD = 100;
+  constexpr size_t BINARY_SEARCH_THRESHOLD = 100;
```
`kBinarySearchThreshold`

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -30,18 +31,24 @@ ProjectAsArrowArray(const T* buf, const size_t len) noexcept {
   using iterator = node_iterator;
+};
 
+class KATANA_EXPORT EdgeShuffleTopology;
```
EdgeShuffle or EdgeSorted?

amberhassaan replied:
EdgeShuffle because Transpose is a form of shuffling edges, so I wanted to pick a more general name that "sorted".

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -88,44 +93,34 @@ class KATANA_EXPORT GraphTopology {
 
   // Edge accessors
```
Drop

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -136,6 +131,23 @@ class KATANA_EXPORT GraphTopology {
 
+  ///@param node node to get degree for
+  ///@returns Degree of node N
+  size_t degree(Node node) const noexcept { return edges(node).size(); }
```
Is `edges().size()` more expensive than difference in `adj_indices_`?

amberhassaan replied:
I don't think so, because, with inlining the compiler should be able to reduce iterator subtraction to subtraction between 2 numbers (counting iterator is after all, a thin wrapper).

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +852,513 @@ CreateSymmetricGraph(PropertyGraph* pg);
 CreateTransposeGraphTopology(const GraphTopology& topology);
 
+class KATANA_EXPORT EdgeTypeIndex : public GraphTopologyTypes {
+  using EdgeTypeIntern = PropertyGraph::TypeSetID;
```
Internal?

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +852,513 @@ CreateSymmetricGraph(PropertyGraph* pg);
+  using EdgeTypeIntern = PropertyGraph::TypeSetID;
+  // map an integer id to each unique edge edge_type in the graph
+  using EdgeTypeIDToIndexMap = std::unordered_map<EdgeTypeIntern, uint32_t>;
+  // each index refers to a unique edge edge_type in the graph
```
Doxygen comments?

nit: Prefer to say that this is a map as well because that's the abstraction.

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +852,513 @@ CreateSymmetricGraph(PropertyGraph* pg);
+  using EdgeIndexToTypeIDMap = std::vector<EdgeTypeIntern>;
+
+public:
+  using EdgeTypeID = EdgeTypeIntern;
```
I didn't understand the need for two types?

amberhassaan replied:
I wanted to keep the two map typedefs private but wanted to expose EdgeTypeIDRange.

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +852,513 @@ CreateSymmetricGraph(PropertyGraph* pg);
+
+  EdgeTypeID GetType(uint32_t index) const noexcept {
+    KATANA_LOG_ASSERT(index < num_unique_edge_types_);
+    KATANA_LOG_ASSERT(size_t(index) < edge_index_to_type_map_.size());
```
debug asserts?

Also, is `size_t{index}` same as `size_t(index)`?

amberhassaan replied:
for `size_t` there's no difference. For a class type, some object that is not of type `T` (let's say obj is of type X), and you write  `T(obj)`, that can mean (1) calling a type conversion operator `operator T()` on obj (in class X) (2) Calling a conversion constructor defined in class `T`. If either of those are defined.

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +852,513 @@ CreateSymmetricGraph(PropertyGraph* pg);
+  }
+
+  uint32_t GetIndex(const EdgeTypeID& edge_type) const noexcept {
+    KATANA_LOG_ASSERT(edge_type_to_index_map_.count(edge_type) > 0);
```
Debug assert?

amberhassaan replied:
good catch.

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +852,513 @@ CreateSymmetricGraph(PropertyGraph* pg);
+
+  uint32_t GetIndex(const EdgeTypeID& edge_type) const noexcept {
+    KATANA_LOG_ASSERT(edge_type_to_index_map_.count(edge_type) > 0);
+    // return edge_type_to_index_map_[edge_type];
```
Drop

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +852,513 @@ CreateSymmetricGraph(PropertyGraph* pg);
+  uint32_t GetIndex(const EdgeTypeID& edge_type) const noexcept {
+    KATANA_LOG_ASSERT(edge_type_to_index_map_.count(edge_type) > 0);
+    // return edge_type_to_index_map_[edge_type];
+    return edge_type_to_index_map_.find(edge_type)->second;
```
`at` instead of `find`?

amberhassaan replied:
Turns out unordered_maps `operator[]` doesn't like being called from a `const` context. But, `at()` has both a `const` and non-const version.

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +852,513 @@ CreateSymmetricGraph(PropertyGraph* pg);
+
+  EdgeTypeIDToIndexMap edge_type_to_index_map_;
+  EdgeIndexToTypeIDMap edge_index_to_type_map_;
+  uint32_t num_unique_edge_types_ = 0u;
```
Is this different than `edge_type_to_index_map_.size()`? Then please document

amberhassaan replied:
Actually, it isn't. I mindlessly lifted it from old code. I think it can be eliminated.

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +852,513 @@ CreateSymmetricGraph(PropertyGraph* pg);
+    return orig_edge_ids_[eid];
+  }
+
+  void SortEdgesByTypeThenDest() noexcept;
```
nit: I'd prefer `SortedEdgeTopology` or something of that sort instead of shuffle.

amberhassaan replied:
My current thinking is to use shuffle as umbrella term for both sorting. We could argue that things like transpose are a sorting afterall. So let me think about it and may be I'll rename it to sorted in the next PR.

roshandathathri replied:
I didn't realize that `Shuffle` includes transpose. Transpose has iteration over in-edge but sorting doesn't, so how/why are they under the same class?

amberhassaan replied:
Think of (going forward as well, for next PR) `Shuffle` as a building block for optional topologies. Transpose is one way to shuffle (or sort) edges. 
Secondly, the concept of `in-edges` only applies to bi-directional topologies. Transpose as a stand-alone doesn't support `in_edges()`, it just supports `edges()`, because it's a building block to something else (e.g., a bidirectional topology). Say I take GraphTopology and transpose it somehow, it's `edges(src)` method will start return `in_edges()`.

roshandathathri replied:
Ok. That makes sense. Please document it in the code in a subsequent PR.

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +852,513 @@ CreateSymmetricGraph(PropertyGraph* pg);
+  /// If not found, returns nothing.
+  // TODO(amber): Assess the usefulness of this method. This method cannot return
+  // edges of all types. Only the first found type. We should however support
+  // find_edges(src, dst) or find_edge(src, dst) that doesn't care about edge type
```
Agreed

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +852,513 @@ CreateSymmetricGraph(PropertyGraph* pg);
+    // EdgeTypeIndex edge_type_index = EdgeTypeIndex::Make(pg);
+    // OutTopoBase out_topo = OutTopoBase::MakeFromDefaultTopology(pg, &edge_type_index);
+    // EdgeTypeAwareTopology in_topo = EdgeTypeAwareTopology::MakeFromTransposeTopology(pg, &edge_type_index);
+    // return EdgeTypeAwareBiDirTopology{std::move(edge_type_index), std::move(out_topo), std::move(in_topo)};
```
Drop?

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -842,6 +852,513 @@ CreateSymmetricGraph(PropertyGraph* pg);
+      return MakeStandardRange<edge_iterator>(Edge{0}, Edge{0});
+    }
+
+    return OutTopoBase::FindAllEdgesSingleType(src, dst);
```
It would be good to choose b/w out and in-edges search. The problem with the latter is that we can need to an easy way to convert in_edges_range to edges_range. Please add a TODO.

roshandathathri commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -906,30 +913,24 @@ katana::GraphTopology::Edge
+  auto e_range = topo.edges(src);
 
-  constexpr int BINARY_SEARCH_THRESHOLD = 100;
+  constexpr size_t kBinarySearchThreshold = 64;
```
+1 on the threshold

roshandathathri commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -906,30 +913,24 @@ katana::GraphTopology::Edge
-    auto iter = std::find_if(e_beg, e_end, [&](const GraphTopology::Edge& e) {
-      return topo.edge_dest(e) == dst;
-    });
+  if (e_range.size() < kBinarySearchThreshold) {
```
`<=`?

roshandathathri commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -1168,3 +1166,245 @@ katana::CreateTransposeGraphTopology(const GraphTopology& topology) {
+  }
+
+  // TODO(amber): introduce a per-thread-container type that frees memory
+  // correctly
```
@l-hoang This is important for you to know.

roshandathathri commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -1168,3 +1166,245 @@ katana::CreateTransposeGraphTopology(const GraphTopology& topology) {
+          orig_edge_ids[e_new] = e;
+        }
+      },
+      katana::no_stats());
```
`katana::steal()`

roshandathathri reviewed (approved):
LGTM. Thanks!

amberhassaan committed:
minor fixes suggested by Roshan's review

amberhassaan commented:
@roshandathathri : addressed all your latest comments. Gonna squash and merge.
